### PR TITLE
Template Matching Method from Paninski Lab

### PR DESCRIPTION
**Pull request by pauladkisson**
# Overview
This new template matching method implements the algorithm designed by the Paninski Lab, which was originally based on Yet Another Spike Sorter (YASS), but subsequently diverged into its own algorithm.  The crux of the method is that spike templates are upsampled at super-resolution (>fs) and scaled in amplitude to better match individual spikes. In reference to this property, I have named the algorithm WobbleMatch because it 'wobbles' the spikes in time and amplitude. I'm open to alternative names if anyone wants to come up with a better one.

## Porting into Spikeinterface
As much as possible, I have followed the existing structure in the template matching module.  Everything runs from the `find_spikes_from_templates()` function just like the other template matching methods.  However, instead of passing in the waveform extractor, I use the templates directly as a parameter in method_kwargs.

## Testing
I have been regularly testing both the original MatchPursuitObjectiveUpsample based implementation and WobbleMatch with 10s of a MEArec recording (32 channels, 7 units).  The algorithm performs nearly perfectly on this toy recording, so I have also added a challenging condition by downsampling the recording up to 10x, which gradually decreases performance and makes the upsampling process more relevant to the matching.  Testing occurs in the `test_wobble_match.py` script. I'm not sure how much of the testing (if any) should be added to the `test_template_matching.py` script.  It may be a bit too heavy(?), but I'm not sure what our philosophy is with testing, so I could use some guidance there.

## Changes
To the best of my knowledge (and tested by various parameter combinations), my implementation is 
functionally identical to the existing one in the MatchPursuitObjectiveUpsample class found [here](https://github.com/cwindolf/spike-psvae/blob/main/spike_psvae/deconvolve.py).  The primary changes have been clarifying many of the variable and method names, and dramatically reorganizing the code. Also, I have identified several aspects of the algorithm that I intend to alter after developing more comprehensive benchmarking -- these changes are identified with
`# TODO` flags.

## Speed
I scrapped the multiprocessing in the initial computations (pairwise convolutions of templates, etc.), and it's possible that I induced some inefficiencies while renaming/reorganizing so WobbleMatch might see a mild slow-down relative to MatchPursuitObjectiveUpsample.  I intend to make a full assessment of speed (profiling different # of units, recording sizes, # of channels, etc.) when the implementation is a bit more mature (after implementing performance-based changes).

### CODE REVIEW

**Comment by pauladkisson** [comment]
On second thought, this PR is already quite large, and I think waiting for more changes will just add more burden on the reviewers.  So, if you wouldn't mind taking a look @alejoe91 @samuelgarcia, let's just merge this version as is (after responding to any feedback).  Then, I can make a separate PR after I implement future improvements.

**Comment by alejoe91** [comment]
Ok! you can start working on new features using this PR as base ;) give us a few days to review

**Comment by h-mayorquin** [comment]
@pauladkisson 
Any chance you could just move the code from here to main? That is, just create a new PR where this changes are done in main instead.

This is a

**Comment by pauladkisson** [comment]
I certainly can, but I'm already working on new stuff (speed-up) on a new branch.  @alejoe91 @samuelgarcia any chance this gets merged before Friday?

File: spikeinterface/sortingcomponents/matching/wobble.py
Comment (h-mayorquin) [inline spikeinterface/sortingcomponents/matching/wobble.py:467]: Does this line modifies the variable traces? 

I am confused as to how is this modified across the loop. Is `find_peaks` or `substract_spike_trains` implicitly modifying it? If so, is traces an attribute? 

I kind of feel that this functions should return something.

**Review by h-mayorquin — COMMENTED** [review]
Hi, I started reading this. I added some comments (which are more my confusions more than anything else). 

Take this as feedback of how someone who does not understand the method will read the code to try to understand it. Maybe this will be useful to improve it.

File: spikeinterface/sortingcomponents/matching/wobble.py
Comment (h-mayorquin) [inline spikeinterface/sortingcomponents/matching/wobble.py:69]: `num_frames` is what we use in core.

File: spikeinterface/sortingcomponents/matching/wobble.py
Comment (h-mayorquin) [inline spikeinterface/sortingcomponents/matching/wobble.py:584]: `calculate_high_resolution_shift`? Iit is better to have functions as verbs

File: spikeinterface/sortingcomponents/matching/wobble.py
Comment (h-mayorquin) [inline spikeinterface/sortingcomponents/matching/wobble.py:823]: `convolve_templates`?

File: spikeinterface/sortingcomponents/matching/wobble.py
Comment (pauladkisson) [reply spikeinterface/sortingcomponents/matching/wobble.py]: Yeah, this dataclass ended up being kind of a catch-all for the stuff that didn't fit into the other dataclasses.  I changed it to only include the template data: compressed_templates, norm, and pairwise_convolution, and decided to just pass around the objective and objective_normalized variables directly.  It's a bit more verbose, but I think the class structure is clearer now.  Lmk if it's still confusing.

File: spikeinterface/sortingcomponents/matching/wobble.py
Comment (pauladkisson) [reply spikeinterface/sortingcomponents/matching/wobble.py]: ``run_array`` was a legacy name from the spike_psvae repo.  I decided to just unpack the whole function into `main_function`, which I think works pretty well.

The scalings and distance metric are useful properties for each spike, so I don't want to just get rid of them.  I think we should add them as attributes to the ``spikes`` output array, but it could be nice to discuss.

File: spikeinterface/sortingcomponents/matching/wobble.py
Comment (pauladkisson) [reply spikeinterface/sortingcomponents/matching/wobble.py]: legacy function name --> agree

File: spikeinterface/sortingcomponents/matching/wobble.py
Comment (pauladkisson) [reply spikeinterface/sortingcomponents/matching/wobble.py]: legacy function name --> agree

File: spikeinterface/sortingcomponents/matching/wobble.py
Comment (pauladkisson) [reply spikeinterface/sortingcomponents/matching/wobble.py]: The traces are not modified during the course of this algorithm.  Instead, it subtracts spikes from the objective itself (which is really just a convolution of the traces with the templates).  

The templates are also not altered during the course of template matching in this alg (which I think is for the best tbh).

So, the only 2 things that get updated every iteration are the objective (as well as the normalized version) and the spike train.

I hope that clears things up, lmk if you have more questions.

**Comment by pauladkisson** [comment]
Thank you @h-mayorquin for the feedback -- it was very helpful.  @alejoe91 @samuelgarcia @h-mayorquin please let me know if there are any other changes you would like to see.

**Review by h-mayorquin — COMMENTED** [review]
Thanks for the previous changes. The code is easier to read now.

Here is some more feedback after a second reading.

Two general comments:
I think that we should start using consistent name for indices. @alejoe91 decided that he wanted to use indices which is what numpy is using in their routine and documentation:

https://numpy.org/doc/stable/reference/arrays.indexing.html

(Altought, even them are 100 consistent)

You have both idx, and ids in your code.

Concerning testing I have two big suggestions:
1) We should test the general algorithm (and I think this applies for other sorting component methods) with a recording that has a different sampling rate. 
2) The methods that you have as stand-alone `compute_scale_amplitudes` and `compute_objective` for example should be tested on their own. I think this is important because now you will start to try to improve them (vectorize `compute_objective` for example) and having a small toy example where you know exactly what to expect will save you from introducing errors as you optimize your code. Concretely, I suggest you buy a small toy example where you know the answer or the answer can be built in a dumb super easy to read way and test the different iterations of your model against it. This will save time and headaches to you and whoever imporves this and also will serve as documentation of what the method does.

File: src/spikeinterface/sortingcomponents/matching/wobble.py
Comment (pauladkisson) [reply src/spikeinterface/sortingcomponents/matching/wobble.py:33]: I absolutely agree with this concept in principle, and I was actually already planning on making this change in a subsequent version.  The reason I've waited is that changing it from frames to ms loses the exact equivalence between my implementation and the one in spike-psvae (which is very useful for debugging).  With that in mind, I am holding off on any changes that actually alter functionality until I have a benchmarking suite established to give me some confidence that the changes are appropriate.  So, my preference is to wait on this, but I have this one (among other planned changes) in a list, so I won't forget them.

File: src/spikeinterface/sortingcomponents/matching/wobble.py
Comment (pauladkisson) [reply src/spikeinterface/sortingcomponents/matching/wobble.py]: Thanks for the suggestion! __post_init__ works perfectly in this case.

File: src/spikeinterface/sortingcomponents/matching/wobble.py
Comment (pauladkisson) [reply src/spikeinterface/sortingcomponents/matching/wobble.py]: Agree --> I changed the other aggregate_x functions to be alternative constructors [from_parameters_and_templates() ], which I think solves the organizational issue that you were pointing out.

File: src/spikeinterface/sortingcomponents/matching/wobble.py
Comment (pauladkisson) [reply src/spikeinterface/sortingcomponents/matching/wobble.py:336]: In the current implementation it is, but this is necessary to ensure equivalence with the spike-psvae version (see [previous comment](https://github.com/SpikeInterface/spikeinterface/pull/1448#discussion_r1160898160)).

File: src/spikeinterface/sortingcomponents/matching/wobble.py
Comment (h-mayorquin) [inline src/spikeinterface/sortingcomponents/matching/wobble.py:849]: This function is also doing two things. Computing and normalizing. I think they should be separated.

File: src/spikeinterface/sortingcomponents/tests/test_wobble.py
Comment (h-mayorquin) [inline src/spikeinterface/sortingcomponents/tests/test_wobble.py:1]: Test should have deterministic behavior. You should use the `rng = default_rng(seed=your_seed_here)`  and use that to generate data.

File: src/spikeinterface/sortingcomponents/tests/test_wobble.py
Comment (h-mayorquin) [inline src/spikeinterface/sortingcomponents/tests/test_wobble.py:26]: What I was proposing before is for you to use copy the `compute_objective` method that you have now written as loops (it is just simple and readable). Then, when you start with newer implementations that vectorize the loops or do more fnacy things then will have to agree this function here. This will both serve as a test and documentation as usuallly the code becomes less clear once you start vectorizing.

File: src/spikeinterface/sortingcomponents/matching/wobble.py
Comment (h-mayorquin) [inline src/spikeinterface/sortingcomponents/matching/wobble.py:281]: Question: is it necessary to keep this as a tuple?
Can you keep each of the attributes exposed as attributes (spatial, temporal, etc).

**Review by h-mayorquin — COMMENTED** [review]
More feedback.

Also, what is the biggest difficulty right now to use Spikeinterface's sparsity instead of the one used here?

File: src/spikeinterface/sortingcomponents/matching/wobble.py
Comment (pauladkisson) [reply src/spikeinterface/sortingcomponents/matching/wobble.py:33]: See #1514

**Comment by pauladkisson** [comment]
> More feedback.
> 
> Also, general question: I notice that you are using an ad-hoc version of sparsity here for this class. What is the biggest difficulty right now to use Spikeinterface's sparsity instead of the one used here?

The only issue is that it would lose the exact equivalence to spike-psvae's version

File: src/spikeinterface/sortingcomponents/matching/wobble.py
Comment (pauladkisson) [reply src/spikeinterface/sortingcomponents/matching/wobble.py:312]: I mean internal consistency throughout the module.

File: src/spikeinterface/sortingcomponents/matching/wobble.py
Comment (pauladkisson) [reply src/spikeinterface/sortingcomponents/matching/wobble.py:24]: Honestly, not sure why conv was added (legacy code)...I think `approx_rank` should do just fine.

File: src/spikeinterface/sortingcomponents/matching/wobble.py
Comment (pauladkisson) [reply src/spikeinterface/sortingcomponents/matching/wobble.py:300]: I think it will be easier to read (and to maintain) if the overarching algorithm description is all in one place.

File: src/spikeinterface/sortingcomponents/matching/wobble.py
Comment (pauladkisson) [reply src/spikeinterface/sortingcomponents/matching/wobble.py:352]: Yeah it was...probably not the way we want it, so I adjusted the organization a bit

File: src/spikeinterface/sortingcomponents/matching/wobble.py
Comment (pauladkisson) [reply src/spikeinterface/sortingcomponents/matching/wobble.py:221]: agree --> fixed

File: src/spikeinterface/sortingcomponents/matching/wobble.py
Comment (h-mayorquin) [reply src/spikeinterface/sortingcomponents/matching/wobble.py:732]: I see. I was -wrongly- assuming that the rightmost matrix in the SVD decomposition should be conjugate, but now I see that numpy's algorithm is already returning the conjugate. 

I would like to suggest changing the initialization reverse with filp to view version that operates in place (that is, use the reversed view in the `compute_objective` and `convolve_templates` functions). Using the reversed sliced notation (see screenshoot) avoids creating and internal copy and can be done in the specific lines where that is required without extra cost. This should not only be faster but also will happen to avoid reading mistakes (i.e. somebody forgetting that the temporal component is reversed in time).

image

File: src/spikeinterface/sortingcomponents/matching/wobble.py
Comment (h-mayorquin) [reply src/spikeinterface/sortingcomponents/matching/wobble.py:258]: Makes sense.

File: src/spikeinterface/sortingcomponents/matching/wobble.py
Comment (h-mayorquin) [reply src/spikeinterface/sortingcomponents/matching/wobble.py]: Thanks.

File: src/spikeinterface/sortingcomponents/matching/wobble.py
Comment (h-mayorquin) [reply src/spikeinterface/sortingcomponents/matching/wobble.py]: So now the `spatially_mask_templates` is not needed anymore?

Comment (h-mayorquin) [reply src/spikeinterface/sortingcomponents/matching/wobble.py]: Forget about it, I now see that you are using the where trick in initialize.

File: src/spikeinterface/sortingcomponents/matching/wobble.py
Comment (h-mayorquin) [reply src/spikeinterface/sortingcomponents/matching/wobble.py]: > > More feedback.
> > Also, general question: I notice that you are using an ad-hoc version of sparsity here for this class. What is the biggest difficulty right now to use Spikeinterface's sparsity instead of the one used here?
> 
> The only issue is that it would lose the exact equivalence to spike-psvae's version. But, I've listed it as a TODO in #1514

Thanks!

File: src/spikeinterface/sortingcomponents/matching/wobble.py
Comment (h-mayorquin) [reply src/spikeinterface/sortingcomponents/matching/wobble.py]: Thanks for changing the attribute. Maybe it also makes sense to change the name of the function to `calculate_square_norm`  (or normed squared as you prefer).

**Comment by h-mayorquin** [comment]
@alejoe91 
I have not review yet the `subtract_spike_train` which is a central part of the code that I don't understand yet. That said, if you feel comfortably with that I am fine. What I would really like to see addressed are the comments about making the tests deterministic though.

**Comment by alejoe91** [comment]
@h-mayorquin yes I agree that tests should be deterministic. @pauladkisson can you take care of that?

File: src/spikeinterface/sortingcomponents/matching/wobble.py
Comment (pauladkisson) [reply src/spikeinterface/sortingcomponents/matching/wobble.py:246]: all of the components of the compressed templates almost always are needed together, so it has been easier to just pass them around together in a tuple.  But, just now, I also exposed them as attributes directly in case one might need to access only one at a time for some reason (it also allows more detailed documentation).

File: src/spikeinterface/sortingcomponents/tests/test_wobble.py
Comment (pauladkisson) [reply src/spikeinterface/sortingcomponents/tests/test_wobble.py]: Oh, I see.  Thanks for the clarification, I've added that now.

**Comment by pauladkisson** [comment]
> 1. We should test the general algorithm (and I think this applies for other sorting component methods) with a recording that has a different sampling rate.

Agree --> see #1521

**Comment by pauladkisson** [comment]
Ok, I've responded to all feedback and written tests for each stand-alone function in wobble.py.  Since this is going to be extended very soon with another PR, I think we are ready to merge.  In the subsequent PR I will have more freedom to deviate from the spike-psvae version when it makes sense to do so. 
 Although I want to be careful about not making it worse -- hence the need for good benchmarks.

**Review by alejoe91 — APPROVED** [review]
@pauladkisson looks great! Thanks for putting so much effort in documentation, it makes reading and understanding much more straightforward!

**Comment by h-mayorquin** [comment]
Thanks for bearing with me as I learned about the method through the code. Also, thanks for adding the tests in the specific manner that I requsted. I think this will save us some hedaches later.

The overall code looks very nice and, as you mentioned, we can make it better after we share this more with Paninski. Congratulations for your first big merge to spikeinterface!

**Pull request merged**
